### PR TITLE
Better handling of error in inherents logic.

### PR DESCRIPTION
**Pull request by gui1117**
Currently the code generated by `construct_runtime` for the check of inherent have a `expect` with a quite light proof.

In case multiple non fatal error are reported for an inherent the proof is not accurate.
This code makes a more explicit handling of the error and handle this case more appropriately.

IMO we should modify more deeply the inherents logic so that `CheckInherentResult` could handle multiple non fatal error for an inherent. Or only allow one error by inherent. And maybe remove non fatal error if it is considered superfluous.

But prior to a more consequent refactor to be discussed I think it is better to have an accurate handling of the result from `put_error`.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3136154

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3136140

**Review by bkchr — APPROVED** [review]

**Review by lexnv — APPROVED** [review]

File: frame/support/procedural/src/construct_runtime/expand/inherent.rs [hunk 1]
```diff
@@ -110,9 +138,9 @@ pub fn expand_outer_inherent(
 								if #pallet_names::is_inherent(call) {
 									is_inherent = true;
 									if let Err(e) = #pallet_names::check_inherent(call, self) {
-										result.put_error(
+										handle_put_error_result(result.put_error(
```
Comment (kianenigma) [inline frame/support/procedural/src/construct_runtime/expand/inherent.rs:141]: I would suggest that we move the `if fata_error { return }` part also to a helper function as this is the part that is still important and ensures we bail after seeing one error.

File: frame/support/procedural/src/construct_runtime/expand/inherent.rs [hunk 2]
```diff
@@ -90,9 +90,37 @@ pub fn expand_outer_inherent(
...
+								id
+							),
+						Err(Error::FatalErrorReported) =>
+							log::error!(
+								target: LOG_TARGET,
+								"Fatal error already reported, unexpected considering there is \
+								only one fatal error",
+							),
+						Err(Error::DecodingFailed(_, _)) | Err(Error::Application(_)) =>
+							log::error!(
```
Comment (kianenigma) [inline frame/support/procedural/src/construct_runtime/expand/inherent.rs:117]: Could use `defensive` macro here instead, as we still know these should (almost) never happen.

**Review by kianenigma — APPROVED** [review]

File: frame/support/procedural/src/construct_runtime/expand/inherent.rs (hunk 2 above)
Comment (thiolliere) [reply frame/support/procedural/src/construct_runtime/expand/inherent.rs:117]: but `defensive` doesn't support renaming of `frame_support` contrary to `construct_runtime`.

(by renaming mean if somebody rename in its Cargo.toml like this:
```
s = { version = "*", package = "frame-support" }
```
)

File: frame/support/procedural/src/construct_runtime/expand/inherent.rs (hunk 1 above)
Comment (thiolliere) [reply frame/support/procedural/src/construct_runtime/expand/inherent.rs:141]: I'm not exactly sure how to factorise it out. something like this?
```diff
@@ -138,10 +138,12 @@ pub fn expand_outer_inherent(
                                                                if #pallet_names::is_inherent(call) {
                                                                        is_inherent = true;
                                                                        if let Err(e) = #pallet_names::check_inherent(call, self) {
-                                                                               handle_put_error_result(result.put_error(
-                                                                                       #pallet_names::INHERENT_IDENTIFIER, &e
+                                                                               let is_fatal = put_error_and_is_error_fatal(
+                                                                                       result,
+                                                                                       #pallet_names::INHERENT_IDENTIFIER,
+                                                                                       &e
                                                                                ));
-                                                                               if e.is_fatal_error() {
+                                                                               if is_fatal {
                                                                                        return result;
                                                                                }
                                                                        }
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3150536

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3150550

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3150619

File: frame/support/procedural/src/construct_runtime/expand/inherent.rs [hunk 3]
```diff
@@ -90,9 +90,37 @@ pub fn expand_outer_inherent(
...
+								error is ignored",
+								id
+							),
+						Err(Error::FatalErrorReported) =>
+							log::error!(
+								target: LOG_TARGET,
+								"Fatal error already reported, unexpected considering there is \
+								only one fatal error",
+							),
+						Err(_) =>
```
Comment (thiolliere) [inline frame/support/procedural/src/construct_runtime/expand/inherent.rs:116]: error variant are different when sp-inherents is compiled with `std` or not.
but some test crate use `construct_runtime` without a `std` feature. So there is no way to know what is the actual error type.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3150805

**Pull request merged**